File: src/main/java/ru/spb/se/contexthelper/context/declr/DeclarationsContextExtractor.java
package ru.spb.se.contexthelper.context.declr;

import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiElement;
import com.intellij.psi.ResolveState;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class DeclarationsContextExtractor {
  @NotNull
  public DeclarationsContext extractContextFor(@NotNull PsiElement psiElement) {
    ContextPsiScopeProcessor scopeProcessor = new ContextPsiScopeProcessor();
    PsiElement currentPsiElement = psiElement;
    PsiElement previousPsiElement = psiElement;
    while (!(currentPsiElement instanceof PsiDirectory)) {
      currentPsiElement.processDeclarations(
          scopeProcessor, ResolveState.initial(), previousPsiElement, psiElement);
      previousPsiElement = currentPsiElement;
      currentPsiElement = currentPsiElement.getParent();
    }
    List<Declaration> declarations = scopeProcessor.getDeclarations();
    return new DeclarationsContext(declarations);
  }
}

File: src/main/java/ru/spb/se/contexthelper/context/declr/DeclarationsContextExtractor.kt
package ru.spb.se.contexthelper.context.declr

import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiElement
import com.intellij.psi.ResolveState

/** Extracts all available declarations up to the given [PsiElement]. */
class DeclarationsContextExtractor(private val psiElement: PsiElement) {
    private val scopeProcessor = ContextPsiScopeProcessor()
    val context: DeclarationsContext

    init {
        extractContextFrom(psiElement, null)
        val declarations = scopeProcessor.declarations
        context = DeclarationsContext(declarations)
    }

    private fun extractContextFrom(currentPsiElement: PsiElement, lastParent: PsiElement?) {
        if (currentPsiElement !is PsiDirectory) {
            currentPsiElement.processDeclarations(
                scopeProcessor, ResolveState.initial(), lastParent, psiElement)
            scopeProcessor.upParentCounter()
            extractContextFrom(currentPsiElement.parent, currentPsiElement)
        }
    }
}